http-common/src/server.rs

```diff
@@ -0,0 +1,319 @@
+														}).to_http_response()),
+													};
+
+													let body: <$route as http_common::server::Route>::DeleteBody = match serde_json::from_slice(&body) {
+														Ok(body) => body,
+														Err(err) => return Ok((http_common::server::Error {
+															status_code: http::StatusCode::UNPROCESSABLE_ENTITY,
+															message: http_common::server::error_to_message(&err).into(),
+														}).to_http_response()),
+													};
+
+													Some(body)
+												}
+												else {
+													None
+												}
+											};
+
+											let (status_code, response) = match <$route as http_common::server::Route>::delete(route, body).await {
+												Ok(result) => result,
+												Err(err) => return Ok(err.to_http_response()),
+											};
+											http_common::server::json_response(status_code, response.as_ref())
+										},
+
+										http::Method::GET => {
+											let (status_code, response) = match <$route as http_common::server::Route>::get(route).await {
+												Ok(result) => result,
+												Err(err) => return Ok(err.to_http_response()),
+											};
+											http_common::server::json_response(status_code, Some(&response))
+										},
+
+										http::Method::POST => {
+											let body = {
+												let content_type = headers.get(hyper::header::CONTENT_TYPE).and_then(|value| value.to_str().ok());
+												if content_type.as_deref() == Some("application/json") {
+													let body = match hyper::body::to_bytes(body).await {
+														Ok(body) => body,
+														Err(err) => return Ok((http_common::server::Error {
+															status_code: http::StatusCode::BAD_REQUEST,
+															message: http_common::server::error_to_message(&err).into(),
+														}).to_http_response()),
+													};
+
+													let body: <$route as http_common::server::Route>::PostBody = match serde_json::from_slice(&body) {
+														Ok(body) => body,
+														Err(err) => return Ok((http_common::server::Error {
+															status_code: http::StatusCode::UNPROCESSABLE_ENTITY,
+															message: http_common::server::error_to_message(&err).into(),
+														}).to_http_response()),
+													};
+
+													Some(body)
+												}
+												else {
+													None
+												}
+											};
+
+											let (status_code, response) = match <$route as http_common::server::Route>::post(route, body).await {
+												Ok(result) => result,
+												Err(err) => return Ok(err.to_http_response()),
+											};
+											http_common::server::json_response(status_code, response.as_ref())
+										},
+
+										http::Method::PUT => {
+											let content_type = headers.get(hyper::header::CONTENT_TYPE).and_then(|value| value.to_str().ok());
+											if content_type.as_deref() != Some("application/json") {
+												return Ok((http_common::server::Error {
```

Comment:
Is a JSON response required for all PUT requests per RFC?

Reply:
`s/JSON response/request body/` ?

If so, we don't have any route that has a `PUT` with no body. If we ever have one, we can make it an `Option` just like `POST`'s arm.

Reply:
I am good with that. Thanks.